src/cos.py

```diff
@@ -22,21 +87,154 @@ def __init__(self, charm: ops.CharmBase, state: State):
 
-    def _on_loki_push_api_endpoint_joined(self, event: LokiPushApiEndpointJoined) -> None:
+        self._loki_consumer = LokiPushApiConsumer(
+            charm=charm, relation_name=METRICS_LOGGING_INTEGRATION_NAME
+        )
+        charm.framework.observe(
+            self._loki_consumer.on.loki_push_api_endpoint_joined,
+            self._on_loki_push_api_endpoint_joined,
+        )
+        charm.framework.observe(
+            self._loki_consumer.on.loki_push_api_endpoint_departed,
+            self._on_loki_push_api_endpoint_departed,
+        )
+        charm.on.define_event("promtail_health", PromtailHealthCheckEvent)
+
+        charm.framework.observe(charm.on.promtail_health, self._on_promtail_health)
+
+    def _retrieve_loki_integration_data(self) -> LokiIntegrationData:
+        """Retrieve and validate Loki integration data.
+
+        Returns:
+            The validated integration data.
+        Raises:
+            pydantic.ValidationError: If the integration data is not valid.
+        """
+        loki_endpoints = []
+        for endpoint in self._loki_consumer.loki_endpoints:
+            loki_endpoints.append(LokiEndpoint(url=endpoint.get("url")))
+        logger.debug("Found following Loki endpoints: %s", loki_endpoints)
+
+        relations: list[Relation] = self.charm.model.relations["metrics-logging"]
+        if relations and (relation := relations[0]).app:
+            promtail_binaries_json = json.loads(
+                relation.data[relation.app].get("promtail_binary_zip_url", "{}")
+            )
+            promtail_binaries = {
+                arch: PromtailBinary(**info) for arch, info in promtail_binaries_json.items()
+            }
+            logger.debug("Found following promtail binaries: %s", promtail_binaries)
+        else:
+            promtail_binaries = {}
+
+        return LokiIntegrationData(endpoints=loki_endpoints, promtail_binaries=promtail_binaries)
+
+    def _validate_for_start(self, loki_integration_data: LokiIntegrationData) -> None:
+        """Validate that the Loki integration data is complete to start Promtail.
+
+        Args:
+            loki_integration_data: The Loki integration data to validate.
+
+        Raises:
+            LokiIntegrationDataNotComplete: If the integration data is not complete.
+        """
+        if not loki_integration_data.endpoints:
+            raise LokiIntegrationDataNotCompleteError("No Loki endpoint found.")
+        if not loki_integration_data.promtail_binaries.get("amd64"):
+            raise LokiIntegrationDataNotCompleteError(
+                "No Promtail binary information for amd64 architecture found."
+            )
+
+    def metrics_logging_available(self) -> bool:
+        """Check that the metrics logging integration is set up correctly.
+
+        Returns:
+            True if the integration is established, False otherwise.
+        """
+        return bool(self._retrieve_loki_integration_data().endpoints)
```

Comment:
Should this be included in `CharmState` rather than as a public function here? One of the ways that can be done is to pass the `requires` instances to `CharmState` and to the observer. The reason that this is relevant information about the charm state that needs to be used for the workload

Reply:
What I'm wondering if we should contribute this to the charm library

Reply:
The observer module is no longer required, and this method is now a property of the charm state.